modules/user_accounts/php/my_preferences.class.inc

```diff
@@ -0,0 +1,1134 @@
+        // changes to the permissions are handled (there should not be any
+        // change to the user's permission set since all the checkboxes are
+        // disabled)
+        if (!$this->_isEditingOwnAccount()) {
+            $success = $user->removePermissions();
+
+            // Check for new permissions
+            if (!empty($permIDs)) {
+                foreach ($permIDs as $permID) {
+                    /* if the user didn't have the permission
+                       and the permission is now assigned then insert
+                       insert into the user_account_history as 'I'
+                     */
+                    if (!(in_array($permID, $current_permissionids))) {
+                        $user->insertIntoUserAccountHistory($permID, 'I');
+                        $permissionsAdded[]
+                            = $this->getDescriptionUsingPermID($permID);
+                    }
+                }
+            }
+
+            if (!empty($current_permissionids)) {
+                // Check for permissions that are to be deleted
+                foreach ($current_permissionids as $currentPermID) {
+                    //if the permission existed before and it's removed now///
+                    ///Then insert into the user_account_history as 'D'
+                    if (!in_array($currentPermID, $permIDs)) {
+                        $user->insertIntoUserAccountHistory($currentPermID, 'D');
+                        $permissionsRemoved[]
+                            = $this->getDescriptionUsingPermID($currentPermID);
+                    }
+                }
+            }
+
+            // send the selected supervisors an email
+            // (only if permissions have changed for the user)
+            if (isset($supervisorEmails)) {
+                foreach ($supervisorEmails as $email => $checkValue) {
+                    if (!empty($permissionsAdded) || !empty($permissionsRemoved)) {
+                        if ($checkValue == 'on') {
+                            $msg_data['current_user'] = $editor->getFullname();
+                            $msg_data['study']        = $config->getSetting('title');
+                            $msg_data['realname']     = $values['Real_name'];
+                            $msg_data['username']     = $user->getUsername();
+                            $msg_data['permissions_added']   = $permissionsAdded;
+                            $msg_data['permissions_removed'] = $permissionsRemoved;
+                            \Email::send(
+                                $email,
+                                'permissions_change_notify_supervisor.tpl',
+                                $msg_data
+                            );
+                        }
+                    }
+                }
+            }
+
+            if (!empty($permIDs)) {
+                $user->addPermissions($permIDs);
+            }
+        }
+
+        // send the user an email
+        if (!empty($send)) {
```

Comment:
is this necessary for changing preferences?

Reply:
no